File: code/estin/assignements/maze/web/frontend/src/lib/agents.js
const directions = [
  [0, 1],
  [0, -1],
  [1, 0],
  [-1, 0],
];

function initialMaze(width = 10, height = 10) {
  return {
    state: Array(height)
      .fill()
      .map(() =>
        Array(width)
          .fill()
          .map(() => ({
            visited: false,
            walls: [true, true, true, true],
          }))
      ),
    width: width,
    height: height,
  };
}

function getNeighbors(x, y, maze) {
  const canVisit = [];
  for (const [dx, dy] of directions) {
    if (
      0 <= x + dx &&
      x + dx < maze.height &&
      0 <= y + dy &&
      y + dy < maze.width
    ) {
      if (maze.state[x + dx][y + dy].visited === false) {
        canVisit.push([x + dx, y + dy]);
      }
    }
  }
  return canVisit;
}

function newWalls(currentCell, x, y, adjX, adjY) {
  const walls = currentCell.walls;
  if (x === adjX && y < adjY) {
    // right
    walls[0] = false;
  }
  if (x === adjX && y > adjY) {
    // left
    walls[1] = false;
  }
  if (x < adjX && y === adjY) {
    // down
    walls[2] = false;
  }
  if (x > adjX && y === adjY) {
    // up
    walls[3] = false;
  }
  return walls;
}

function generateMaze(height = 10, width = 10, start = [0, 0]) {
  const maze = initialMaze(width, height);
  const stack = [start];

  const startTime = Date.now();

  while (stack.length > 0) {
    const current = stack.pop();
    const [x, y] = current;
    maze.state[x][y].visited = true;
    const adjs = getNeighbors(x, y, maze);

    if (adjs.length > 0) {
      const adj = adjs[Math.floor(Math.random() * adjs.length)];
      const [adjX, adjY] = adj;
      maze.state[x][y].walls = newWalls(maze.state[x][y], x, y, adjX, adjY);
      maze.state[adjX][adjY].walls = newWalls(
        maze.state[adjX][adjY],
        adjX,
        adjY,
        x,
        y
      );
      stack.push([x, y]);
      stack.push([adjX, adjY]);
    }
  }

  const endTime = Date.now();

  return [maze, (endTime - startTime) / 1000];
}

function getReachableNeighbours(maze, cred) {
  const directions = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ]; // right left down up
  const [i, j] = cred;
  const canVisit = [];

  for (let idx = 0; idx < maze.state[i][j].walls.length; idx++) {
    if (maze.state[i][j].walls[idx] === false) {
      const [x, y] = directions[idx];
      canVisit.push([x + i, y + j]);
    }
  }

  return canVisit;
}

function euclidianDistanceHeuristic(a, b) {
  return Math.sqrt(Math.pow(b[0] - a[0], 2) + Math.pow(b[1] - a[1], 2));
}

function manhattanDistanceHeuristic(a, b) {
  return Math.abs(b[0] - a[0]) + Math.abs(b[1] - a[1]);
}

// Agents

function bfsAgent(maze, start, end, withSteps = false) {
  const visited = new Set();
  const queue = [[start]];

  const steps = []; // has nothing to do with the algorithm, just for visualization
  const startTime = Date.now();

  while (queue.length > 0) {
    const path = queue.shift();
    const node = path[path.length - 1];
    steps.push([...path]);

    const nodeStr = JSON.stringify(node); // for Set comparison

    if (!visited.has(nodeStr)) {
      visited.add(nodeStr);

      if (node[0] === end[0] && node[1] === end[1]) {
        const endTime = Date.now();
        if (!withSteps) {
          return {
            path: path,
            time: (endTime - startTime) / 1000,
            visited: Array.from(visited).map((v) => JSON.parse(v)),
          };
        }
        return {
          path: path,
          time: (endTime - startTime) / 1000,
          visited: Array.from(visited).map((v) => JSON.parse(v)),
          steps: steps,
        };
      }

      const neighbours = getReachableNeighbours(maze, node);
      for (const neighbour of neighbours) {
        const neighbourStr = JSON.stringify(neighbour);
        if (!visited.has(neighbourStr)) {
          const newPath = [...path];
          newPath.push(neighbour);
          queue.push(newPath);
        }
      }
    }
  }

  return null;
}

function dfsAgent(maze, start, end, withSteps = false) {
  const visited = new Set();
  const stack = [[start]];

  const steps = [];
  const startTime = Date.now();

  while (stack.length > 0) {
    const path = stack.pop();
    const node = path[path.length - 1];
    steps.push([...path]);

    const nodeStr = JSON.stringify(node); // for Set comparison

    if (!visited.has(nodeStr)) {
      visited.add(nodeStr);

      if (node[0] === end[0] && node[1] === end[1]) {
        const endTime = Date.now();
        if (!withSteps) {
          return {
            path: path,
            time: (endTime - startTime) / 1000,
            visited: Array.from(visited).map((v) => JSON.parse(v)),
          };
        }
        return {
          path: path,
          time: (endTime - startTime) / 1000,
          visited: Array.from(visited).map((v) => JSON.parse(v)),
          steps: steps,
        };
      }

      const neighbours = getReachableNeighbours(maze, node);
      for (const neighbour of neighbours) {
        const neighbourStr = JSON.stringify(neighbour);
        if (!visited.has(neighbourStr)) {
          const newPath = [...path];
          newPath.push(neighbour);
          stack.push(newPath);
        }
      }
    }
  }

  return null;
}

function aStarAgent(maze, start, end, heuristic, withSteps = false) {
  class PriorityQueue {
    constructor() {
      this.elements = [];
    }

    push(item, priority) {
      this.elements.push({ item, priority });
      this.elements.sort((a, b) => a.priority - b.priority);
    }

    pop() {
      return this.elements.shift().item;
    }

    isEmpty() {
      return this.elements.length === 0;
    }
  }

  const gScore = new Map();
  const openSet = new PriorityQueue();
  const visited = new Set();

  gScore.set(JSON.stringify(start), 0);
  openSet.push([heuristic(start, end), start, [start]], heuristic(start, end));

  const startTime = Date.now();
  const steps = [];

  while (!openSet.isEmpty()) {
    const [f, current, path] = openSet.pop();
    steps.push([...path]);

    const currentStr = JSON.stringify(current);

    if (current[0] === end[0] && current[1] === end[1]) {
      const endTime = Date.now();
      if (!withSteps) {
        return {
          path: path,
          time: (endTime - startTime) / 1000,
          visited: Array.from(visited).map((v) => JSON.parse(v)),
        };
      }
      return {
        path: path,
        time: (endTime - startTime) / 1000,
        visited: Array.from(visited).map((v) => JSON.parse(v)),
        steps: steps,
      };
    }

    if (visited.has(currentStr)) {
      continue;
    }

    visited.add(currentStr);

    for (const neighbor of getReachableNeighbours(maze, current)) {
      const neighborStr = JSON.stringify(neighbor);

      const tentativeG = gScore.get(currentStr) + 1;

      if (!gScore.has(neighborStr) || tentativeG < gScore.get(neighborStr)) {
        gScore.set(neighborStr, tentativeG);

        const h = heuristic(neighbor, end);
        const f = tentativeG + h;
        openSet.push([f, neighbor, [...path, neighbor]], f);
      }
    }
  }

  return null;
}

export {
  bfsAgent,
  dfsAgent,
  generateMaze,
  aStarAgent,
  manhattanDistanceHeuristic,
  euclidianDistanceHeuristic,
};

File: code/estin/assignements/maze/web/api/agents.py
import random
import math
import heapq 
import time

directions = [(0,1),(0,-1),(1,0),(-1,0)]

def initial_maze(width=10,height=10):
    return {
        "state":[[ {"visited":False ,"walls":[True]*4 } for _ in range(width) ] for _ in range(height)] ,
        "width":width,
        "height":height
            }

def get_neighbors(x,y,maze):
    can_visit = []
    for di in directions :
        dx,dy = di
        if  0 <= x + dx < maze["height"] and 0 <= y + dy < maze["width"] : 
            if maze["state"][x + dx][y + dy]["visited"] is False :
                can_visit.append((x+dx,y+dy))
    return  can_visit

def new_walls(current_cell,x,y,adj_x,adj_y):
    walls = current_cell["walls"]
    if x == adj_x and y < adj_y :   #right
        walls[0] = False
    if x == adj_x and y > adj_y :   #left
        walls[1] = False
    if x < adj_x and y == adj_y :   #down
        walls[2] = False
    if x > adj_x and y == adj_y :   #up
        walls[3] = False
    return walls


def generate_maze(height=10,width=10,start=(0,0)):

    maze = initial_maze(height,width)

    stack = [start]

    start_time = time.time()

    while stack :
        current = stack.pop(-1)
        x , y = current
        maze["state"][x][y]["visited"] = True
        adjs = get_neighbors(x,y,maze)
        if adjs :
            adj = random.choice(adjs)
            adj_x,adj_y = adj
            maze["state"][x][y]["walls"] = new_walls(maze["state"][x][y],x,y,adj_x,adj_y)
            maze["state"][adj_x][adj_y]["walls"] = new_walls(maze["state"][adj_x][adj_y],adj_x,adj_y,x,y)
            stack.append((x, y))
            stack.append((adj_x,adj_y))

    end_time = time.time()

    return maze , end_time - start_time

def get_reachable_neighbours(maze,cred):
    directions = [(0,1),(0,-1),(1,0),(-1,0)] #right left down up
    i , j = cred
    can_visit = []
    for idx , val in enumerate(maze["state"][i][j]["walls"]) :
        if val is False :
            x , y = directions[idx]
            can_visit.append((x+i,y+j))
    return  can_visit


def euclidian_distance_heuristic(a,b): 
    return math.sqrt((b[0] - a[0])**2 + (b[1] - a[1])**2 )


def manhattan_distance_heuristic(a,b):
    return  abs(b[0] - a[0]) + abs(b[1] - a[1])


# Agents

def bfs_agent(maze, start, end, with_steps=False):
    visited = set()
    queue = [[start]]  

    steps = [] # has nothing to do with the algorithm, just for visualization
    start_time = time.time()
    
    while queue:
        path = queue.pop(0)
        node = path[-1]
        steps.append(path)
        
        if node not in visited:
            visited.add(node)
            
            if node == end:
                end_time = time.time()
                if not with_steps:
                    return {
                            "path":path ,
                            "time": end_time - start_time ,
                            "visited":visited
                            }
                return {
                        "path":path ,
                        "time": end_time - start_time ,
                        "visited":visited ,
                        "steps":steps
                        }
            
            neighbours = get_reachable_neighbours(maze,node)
            for neighbour in neighbours:
                if neighbour not in visited:
                    new_path = list(path)
                    new_path.append(neighbour)
                    queue.append(new_path)
    
    return None

def dfs_agent(maze, start, end, with_steps=False):
    visited = set()
    stack = [[start]]  

    steps = []
    start_time = time.time()
    
    while stack:
        path = stack.pop()
        node = path[-1]
        steps.append(path)
        
        if node not in visited:
            visited.add(node)
            
            if node == end:
                end_time = time.time()
                if not with_steps:
                    return {
                            "path":path ,
                            "time": end_time - start_time ,
                            "visited":visited
                            }
                return {
                        "path":path ,
                        "time": end_time - start_time ,
                        "visited":visited ,
                        "steps":steps
                        }
            
            neighbours = get_reachable_neighbours(maze,node)
            for neighbour in neighbours:
                if neighbour not in visited:
                    new_path = list(path)
                    new_path.append(neighbour)
                    stack.append(new_path)
    
    return None



def a_star_agent(maze, start, end , heuristic, with_steps=False):

    open_set = [] 
    closed_set = {} 

    heapq.heappush(open_set,(0,start,[start]))
    closed_set[start] = 0

    start_time = time.time()
    steps = []

    while open_set :
        f , current , path = heapq.heappop(open_set)
        steps.append(path)

        if current == end :
            end_time = time.time()
            if not with_steps:
                return {
                        "path":path ,
                        "time": end_time - start_time ,
                        "visited":closed_set.keys()
                        }
            return {
                    "path":path ,
                    "time": end_time - start_time ,
                    "visited":closed_set.keys() ,
                    "steps":steps
                    }

        if current in closed_set and closed_set[current] < f :
            continue

        if (current in closed_set and closed_set[current] > f) or current not in closed_set :
            closed_set[current] = f

        for neighbor in get_reachable_neighbours(maze , current) :
            g = len(path)
            h = heuristic(neighbor,end)
            f = h + g
            heapq.heappush(open_set,(f,neighbor,path + [neighbor]))

    return None
